test/eventbus.test.ts

```diff
@@ -271,6 +271,71 @@ test("map narrows type and pipe enforces", () => {
       (event): event is EventBusRuleInput<t1> => event.detail.type === "one"
     )
     .map((event) => event.detail.one)
-    // should fail compilation if the types don't match
     .pipe(lambda);
 });
+
+test("pipe typesafe", () => {
+  const lambda = Function.fromFunction<string, void>(
+    aws_lambda.Function.fromFunctionArn(stack, "func", "")
+  );
+  const bus = EventBus.default<tt>(stack);
+
+  bus
+    .when(
+      stack,
+      "rule",
+      (event): event is EventBusRuleInput<t1> => event.detail.type === "one"
+    )
+    .map((event) => event.detail) // is object
+    // @ts-expect-error should fail compilation if the types don't match
+    .pipe(lambda); // expects strings
+});
+
+test("pipe typesafe sfn", () => {
+  const sfn = new StepFunction(stack, "machine", (payload: { id: string }) => {
+    return payload.id;
+  });
+  const bus = EventBus.default<tt>(stack);
+
+  bus
+    .when(
+      stack,
+      "rule",
+      (event): event is EventBusRuleInput<t1> => event.detail.type === "one"
+    )
+    .map((event) => ({ id: event.detail.one })) // is object
+    .pipe(sfn); // expects strings
+});
+
+test("pipe typesafe error sfn", () => {
+  const sfn = new StepFunction(stack, "machine", (payload: { id: string }) => {
+    return payload.id;
+  });
+  const bus = EventBus.default<tt>(stack);
+
+  bus
+    .when(
+      stack,
+      "rule",
+      (event): event is EventBusRuleInput<t1> => event.detail.type === "one"
+    )
+    .map((event) => ({ id: event.detail })) // is object
+    // @ts-expect-error
+    .pipe(sfn); // expects strings
+});
+
+test("map cannot pipe to a bus", () => {
+  const bus = EventBus.default<tt>(stack);
+
+  expect(() =>
+    bus
+      .when(
+        stack,
+        "rule",
+        (event): event is EventBusRuleInput<t1> => event.detail.type === "one"
+      )
+      .map((event) => event)
+      // @ts-expect-error
+      .pipe(bus)
+  ).toThrow();
+});
```

Comment:
This is a beautiful test. I love how it has `@ts-expect-error` and `expect().toThrow()`. Total parity between runtime and compile time logic - the types properly describe valid and invalid states.